File: src/types/course.ts
/**
 * Shared types for the LearnX Course ecosystem — the data model that
 * connects Doctor-authored courses to Student enrollment & learning.
 *
 * This is a frontend-only build (see `backend/` for the reference auth
 * architecture), so the "catalog" lives in client state via
 * `useCourseCatalog` and is seeded from `src/data/coursesMock.ts`. The
 * shape is deliberately backend-shaped (status workflow, module/lesson
 * tree, per-course analytics) so it maps cleanly onto a real API later.
 */

export type CourseStatus = 'draft' | 'pending-review' | 'published' | 'archived'

export type CourseType = 'university' | 'public' | 'premium'

export type LessonType = 'video' | 'pdf' | 'notes' | 'quiz' | 'assignment'

export interface LessonResource {
  id: string
  name: string
  kind: 'pdf' | 'docx' | 'ppt' | 'link' | 'dataset'
  sizeLabel: string
}

export interface Lesson {
  id: string
  title: string
  type: LessonType
  durationMinutes?: number
  /** Whether the single demo student ("Alex Chen") has completed this lesson. */
  completed: boolean
  resources: LessonResource[]
}

export interface Module {
  id: string
  title: string
  lessons: Lesson[]
}

export interface CourseAnalytics {
  totalStudents: number
  activeStudents: number
  completionRate: number
  avgWatchTimeMinutes: number
  mostViewedLessonTitle: string
  dropOffLessonTitle: string
  quizAvgScore: number
  strugglingTopic: string
  strugglingPct: number
  aiInsights: string[]
}

export interface Course {
  id: string
  title: string
  description: string
  category: string
  faculty: string
  department: string
  academicLevel: string
  courseType: CourseType
  status: CourseStatus
  color: string
  icon: string
  doctorName: string
  doctorInitials: string
  rating: number
  studentsCount: number
  completionRate: number
  lastUpdated: string
  createdAt: string
  modules: Module[]
  analytics: CourseAnalytics
  /** Student-facing enrollment/progress state (single demo persona: Alex Chen). */
  enrolled: boolean
  saved: boolean
  progressPct: number
  lastLessonTitle: string | null
  lastViewedAt: string | null
  completedAt: string | null
  /**
   * Monetization — only meaningful for `courseType === 'premium'`.
   * `priceUsd` is the real-money price a doctor sets when publishing a
   * premium course; `allowXpRedemption`/`xpPrice` let a doctor opt the
   * course into the Reward Store's "pay with XP instead of money"
   * exchange (spec: "Teachers can choose whether a paid course supports
   * XP redemption"). Both null for non-premium courses.
   */
  priceUsd: number | null
  allowXpRedemption: boolean
  xpPrice: number | null
  /** Real purchase record for a student who redeemed this course via the
   * Reward Store (XP or XP+money) rather than the ordinary Enroll flow. */
  purchasedViaReward: boolean
}

export function totalLessons(course: Course): number {
  return course.modules.reduce((sum, m) => sum + m.lessons.length, 0)
}

export function completedLessons(course: Course): number {
  return course.modules.reduce((sum, m) => sum + m.lessons.filter((l) => l.completed).length, 0)
}

export function remainingLessons(course: Course): number {
  return totalLessons(course) - completedLessons(course)
}

File: backend/app/schemas/course.py
"""Course API schemas — mirror the frontend `src/types/course.ts` shapes."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

CourseStatus = Literal["draft", "pending-review", "published", "archived"]
CourseType = Literal["university", "public", "premium"]
LessonType = Literal["video", "pdf", "notes", "quiz", "assignment"]


class LessonResourceIn(BaseModel):
    name: str
    kind: Literal["pdf", "docx", "ppt", "link", "dataset"]
    sizeLabel: str = ""


class LessonOut(BaseModel):
    id: str
    title: str
    type: LessonType
    durationMinutes: int | None = None
    completed: bool = False
    resources: list = []


class LessonIn(BaseModel):
    title: str
    type: LessonType = "video"
    durationMinutes: int | None = None
    resources: list[LessonResourceIn] = []


class ModuleOut(BaseModel):
    id: str
    title: str
    lessons: list[LessonOut] = []


class ModuleIn(BaseModel):
    title: str


class CourseBase(BaseModel):
    title: str = ""
    description: str = ""
    category: str = ""
    faculty: str = ""
    department: str = ""
    academicLevel: str = ""
    courseType: CourseType = "university"
    status: CourseStatus = "draft"
    color: str = "#2DD4BF"
    icon: str = "📘"
    priceUsd: float | None = None
    allowXpRedemption: bool = False
    xpPrice: int | None = None


class CourseCreate(CourseBase):
    pass


class CourseUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    category: str | None = None
    faculty: str | None = None
    department: str | None = None
    academicLevel: str | None = None
    courseType: CourseType | None = None
    status: CourseStatus | None = None
    color: str | None = None
    icon: str | None = None
    priceUsd: float | None = None
    allowXpRedemption: bool | None = None
    xpPrice: int | None = None


class CourseOut(CourseBase):
    id: str
    doctorName: str = ""
    doctorInitials: str = ""
    rating: float = 4.5
    studentsCount: int = 0
    completionRate: float = 0.0
    createdAt: str = ""
    lastUpdated: str = ""
    modules: list[ModuleOut] = []
    enrolled: bool = False
    saved: bool = False
    progressPct: int = 0
    completedLessonIds: list[str] = []
    purchasedViaReward: bool = False


class EnrollmentToggleOut(BaseModel):
    enrolled: bool
    progressPct: int = 0


class SavedToggleOut(BaseModel):
    saved: bool


class LessonCompleteOut(BaseModel):
    lessonId: str
    completed: bool
    progressPct: int


class RosterStudentCourse(BaseModel):
    course_id: str
    title: str
    progress_pct: int
    enrolled_at: str | None = None
    completed_at: str | None = None


class RosterStudentOut(BaseModel):
    id: str
    email: str
    full_name: str
    avatar_url: str | None = None
    courses: list[RosterStudentCourse] = []
